yt_dlp/extractor/zaiko.py

```diff
@@ -66,7 +66,9 @@ def _real_extract(self, url):
             stream_meta['stream-access']['video_source'], video_id,
             'Downloading player page', headers={'referer': 'https://zaiko.io/'})
         player_meta = self._parse_vue_element_attr('player', player_page, video_id)
-        status = traverse_obj(player_meta, ('initial_event_info', 'status', {str}))
+        initial_event_info = player_meta['initial_event_info']
```

Comment:
The rest of the extractor seemed to be written to be non-fatal all the way down to the expected `raise_no_formats` call. Will making this line fatal have any ill effect? (I'm just asking a question, not requesting a change here)

Reply:
> Will making this line fatal have any ill effect?

Well, I didn't check it carefully. Not all information is extracted from the "initial_event_info" object (we also have "player_meta"), so making it fatal is not ideal.

The design of "fatal vs. non-fatal" always confuses me, btw.

Reply:
> The design of "fatal vs. non-fatal" always confuses me, btw.

When I use the term "fatal" in extractor PR reviews, I mean that code being reviewed could raise an error before `_real_extract` returns the info dict. Metadata extraction should always be non-fatal. Formats extraction can be fatal, but in the case of a livestream we should make every attempt to have it be non-fatal so that we support `--wait-for-video` and extraction of `live_status`/`release_timestamp` etc.

If we are reasonable certain that `player_metadata` always has an `initial_event_info` key, and we actually *want* to raise a KeyError if it doesn't have an `initial_event_info` key (e.g. if this indicates the site has changed and we want a bug report so we can fix it), then the code above is fine as-is.

If there are any currently known cases where `player_metadata` doesn't have an `initial_event_info` key, then we should use `traverse_obj` here so that we can safely return an info dict (and not break `--wait-for-video`)

Reply:
```suggestion
        initial_event_info = traverse_obj(player_meta, ('initial_event_info', {dict})) or {}
```

Reply:
Hey, @bashonly! Thanks for your explanation and fixes.

These days I've been too busy to make changes in my PRs. I'm sorry that I didn't say "_I'll go back when I'm free_". My bad.

Now let's go back to the code. Sometimes I wrote the code "totally non-fatal", and other times I mixed "fatal" and "non-fatal" for no clear reasons. Such code in my other PRs may have been merged with fixes from maintainers, but the thing in my brain is still fuzzy (is it just me? 😟). So, I think the design concept of "fatal vs. non-fatal" could be a part of our contribution guide. Future develops may want to know this before submitting code.

Last year I raised a question about "UserNotLive" and "raise_no_formats()" (https://github.com/yt-dlp/yt-dlp/issues/6981). These two exceptions, like metadata and formats, are output of an extractor. As a coder, personally, I think knowing the common possible output is kind of important. If our guide has content about exceptions, it will be more informative, but looonger.

The code controls the output of an extractor, so I mentioned the old issue. A mismatched key produces a "KeyError", that is OK because we write the code in this way. So:

```
--- Alice: "In what way?"
--- Bob:   "Please see CONTRIBUTING.md."

```

.

Reply:
> These days I've been too busy to make changes in my PRs. I'm sorry that I didn't say "I'll go back when I'm free". My bad.

All good. It's volunteer open source work. It happens

> So, I think the design concept of "fatal vs. non-fatal" could be a part of our contribution guide. Future develops may want to know this before submitting code.

Agreed, the developer section of the contributing guide should probably cover (at least) a few more concepts.

> If our guide has content about exceptions, it will be more informative, but looonger.

This is the problem, though. Where do we stop? Something like "fatal" vs "non-fatal" extraction applies to almost all extractor code, and adding it to the guide would be justified IMO. But I can't say the same for the less commonly applicable subjects like how/when to use UserNotLive vs raise_no_formats. Maybe if someone wants to step up and make/maintain a developer's wiki, then we'd have a place for all of this info. But until then, there are a lot of things that will unfortunately have to remain tribal knowledge with their explanations (if any) buried in the issue/PR tracker.